backends/p4tools/common/lib/symbolic_env.cpp

```diff
@@ -28,7 +28,16 @@ const IR::Expression *SymbolicEnv::get(const IR::StateVariable &var) const {
 bool SymbolicEnv::exists(const IR::StateVariable &var) const { return map.find(var) != map.end(); }
 
 void SymbolicEnv::set(const IR::StateVariable &var, const IR::Expression *value) {
-    map[var] = P4::optimizeExpression(value);
+    const auto *type = value->type;
+    BUG_CHECK(type && !type->is<IR::Type_Unknown>(), "Cannot set value with unspecified type: %1%",
+              value);
+    value = P4::optimizeExpression(value);
+    BUG_CHECK(value->type && !value->type->is<IR::Type_Unknown>(),
+              "The P4 expression optimizer stripped a type of %1% (was %2%)", value, type);
+    BUG_CHECK(type->equiv(*value->type),
```

Comment:
I think it is fair to change the underlying type of the expression. You could imagine something like `bit<8>(0)` ++ `bit<8>(4)` being optimized to `bit<16>(4)`.

Reply:
In that case the `++` would still be `bit<16>`. If an optimizer changed type I think that would have to break type checking.

Reply:
Duh. My bad. Can't think of any other example. Even casts preserve the type.